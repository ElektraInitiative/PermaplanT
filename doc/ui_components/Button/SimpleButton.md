# SimpleButton

## Description

A simple HTML button that may contain other react nodes.

## Props

The props are an extension of React.ButtonHTMLAttributes<HTMLButtonElement>,
so all the props that can be applied to an HTML button can also be applied to this component.

| Name    | Type          | Optional | Description                                                                          |
| ------- | ------------- | -------- | ------------------------------------------------------------------------------------ |
| variant | ButtonVariant | yes      | The variant specifies the look of the button. The options are primary and secondary. |
