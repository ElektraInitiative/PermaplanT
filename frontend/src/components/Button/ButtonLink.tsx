import { MouseEventHandler } from 'react';
import { Link } from 'react-router-dom';

interface ButtonLinkProps {
  /** The title of the button. */
  title: string;
  /** The path to navigate to. */
  to: string;
  /** The className styling of the button. */
  className?: string;
  /** Clickhandler for the button. */
  onClick?: MouseEventHandler<HTMLButtonElement>;
}

/**
 * A styled link that has properties of a button.
 */
const ButtonLink = ({ title, to, className = '', onClick }: ButtonLinkProps) => {
  return (
    <Link
      to={to}
      className={`flex items-center ${className} justify-center space-x-2 active:text-primary-700 dark:active:text-primary-400`}
    >
      <button onClick={onClick}>{title}</button>
    </Link>
  );
};

export default ButtonLink;
