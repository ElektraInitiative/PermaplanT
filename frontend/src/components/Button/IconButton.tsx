interface IconButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** The variant specifies the look of the button. */
  variant?: ButtonVariant;
  /** Active icons (e.g. Grid shown) are rendered in highlighted fashion. */
  renderAsActive?: boolean;
  /** Toolbox icons receive additional styling. */
  isToolboxIcon?: boolean;
}

export enum ButtonVariant {
  primary,
  secondary,
}

const variantStyles = {
  [ButtonVariant.primary]:
    'hover:text-primary-500 dark:hover:text-primary-500 hover:stroke-primary-500 dark:hover:stroke-primary-500 active:stroke-primary-200 dark:active:stroke-primary-200 hover:fill-primary-500 dark:hover:fill-primary-500 active:fill-primary-200 dark:active:fill-primary-200 focus:ring-primary-300 ',
  [ButtonVariant.secondary]:
    'hover:text-secondary-500 dark:hover:text-secondary-500 hover:stroke-secondary-500 dark:hover:stroke-secondary-500 active:stroke-secondary-200 dark:active:stroke-secondary-200 hover:fill-secondary-500 dark:hover:fill-secondary-500 active:fill-secondary-200 dark:active:fill-secondary-200 focus:ring-secondary-300 ',
};

/**
 * A clickable icon.
 * @param props.variant The variant specifies the look of the button.
 * @param props All React props for buttons can be applied.
 */
export default function IconButton({
  variant = ButtonVariant.primary,
  renderAsActive = false,
  isToolboxIcon = false,
  ...props
}: IconButtonProps) {
  const defaultIconStyles =
    'inline-flex h-6 w-6 justify-center rounded-lg items-center text-sm font-medium focus:outline-none focus:ring-1 focus:ring-secondary-100 dark:focus:ring-secondary-500 focus:border-0 dark:focus:border-0 stroke-neutral-800 dark:stroke-neutral-800-dark fill-neutral-800 dark:fill-neutral-800-dark' +
    ' disabled:stroke-neutral-500 dark:disabled:stroke-neutral-500-dark disabled:fill-neutral-500 dark:disabled:fill-neutral-500-dark disabled:cursor-not-allowed' +
    ' ' +
    variantStyles[variant];

  const activeIcon = renderAsActive
    ? 'fill-primary-500 dark:fill-primary-400 stroke-primary-500 dark:stroke-primary-400 border-primary-500 '
    : '';
  const toolboxIcon = isToolboxIcon
    ? 'mx-1 my-2 first-of-type:ml-2 last-of-type:mr-2 h-8 w-8 p-1 border border-neutral-500 hover:bg-neutral-200 dark:hover:bg-neutral-800 dark:hover:stroke-primary-400 active:stroke-primary-400 active:fill-primary-400'
    : '';
  const additionalClasses = props.className ? props.className : '';

  return (
    <button
      type="button"
      {...props}
      className={activeIcon + ' ' + toolboxIcon + ' ' + defaultIconStyles + ' ' + additionalClasses}
    ></button>
  );
}
