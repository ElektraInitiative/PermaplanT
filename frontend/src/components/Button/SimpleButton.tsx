interface SimpleButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** The variant specifies the look of the button. */
  variant?: ButtonVariant;
}

export enum ButtonVariant {
  primaryBase,
  secondaryBase,
  primaryContainer,
  secondaryContainer,
  dangerBase,
}

const variantStyles = {
  [ButtonVariant.primaryBase]:
    'bg-primary-500 dark:bg-primary-300 border-primary-500 hover:bg-primary-600 dark:hover:bg-primary-200 active:bg-primary-900 text-primary-50 dark:text-primary-700',
  [ButtonVariant.secondaryBase]:
    'bg-secondary-500 dark:bg-secondary-300 border-secondary-500 hover:bg-secondary-600 dark:hover:bg-secondary-200 active:bg-secondary-900 text-secondary-50 dark:text-secondary-700',
  [ButtonVariant.primaryContainer]:
    'bg-primary-200 dark:bg-primary-600 border-primary-500 hover:bg-primary-200 dark:hover:bg-primary-600 active:bg-primary-900 text-primary-900 dark:text-primary-200',
  [ButtonVariant.secondaryContainer]:
    'bg-secondary-200 dark:bg-secondary-600 border-secondary-500 hover:bg-secondary-200 dark:hover:bg-secondary-600 active:bg-secondary-900 text-secondary-900 dark:text-secondary-200',
  [ButtonVariant.dangerBase]:
    'bg-red-500 dark:bg-red-300 border-red-500 hover:bg-red-600 dark:hover:bg-red-200 active:bg-red-900 text-red-50 dark:text-red-700',
};

/**
 * A styled HTML button.
 * @param props All React props for buttons can be applied.
 */
export default function SimpleButton({
  variant = ButtonVariant.primaryBase,
  ...props
}: SimpleButtonProps) {
  const className =
    'button disabled:bg-neutral-300 disabled:border-neutral-300 disabled:text-neutral-500 dark:disabled:bg-neutral-300-dark dark:disabled:border-neutral-300-dark dark:disabled:text-neutral-500-dark disabled:cursor-not-allowed focus:outline-none focus:ring-4 focus:ring-blue-300 border ' +
    variantStyles[variant];

  return (
    <button
      type="button"
      {...props}
      className={className + (props.className ? ' ' + props.className : '')}
    ></button>
  );
}
