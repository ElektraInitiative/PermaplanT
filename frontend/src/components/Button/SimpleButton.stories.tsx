import type { Meta, StoryObj } from '@storybook/react';
import SimpleButton, { ButtonVariant } from './SimpleButton';

const meta: Meta<typeof SimpleButton> = {
  title: 'Components/Button/SimpleButton',
  component: SimpleButton,
};

export default meta;

type Story = StoryObj<typeof SimpleButton>;

export const PrimaryBase: Story = {
  args: {
    variant: ButtonVariant.primaryBase,
    children: 'Simple Button',
  },
};

export const PrimaryContainer: Story = {
  args: {
    ...PrimaryBase.args,
    variant: ButtonVariant.primaryContainer,
  },
};

export const SecondaryBase: Story = {
  args: {
    ...PrimaryBase.args,
    variant: ButtonVariant.secondaryBase,
  },
};

export const SecondaryContainer: Story = {
  args: {
    ...PrimaryBase.args,
    variant: ButtonVariant.secondaryContainer,
  },
};
