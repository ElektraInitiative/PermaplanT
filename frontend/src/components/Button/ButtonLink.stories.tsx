import type { Meta, StoryObj } from '@storybook/react';
import { reactRouterDecorator } from '@/utils/stories/react-router-decorators';
import ButtonLink from './ButtonLink';

const meta: Meta<typeof ButtonLink> = {
  title: 'Components/Button/ButtonLink',
  component: ButtonLink,
  decorators: [reactRouterDecorator],
};

export default meta;

type Story = StoryObj<typeof ButtonLink>;

export const Default: Story = {
  args: {
    title: 'Link',
    to: '#',
  },
};
