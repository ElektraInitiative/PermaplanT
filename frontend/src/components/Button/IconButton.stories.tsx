import IconButton, { ButtonVariant } from './IconButton';
import { reactRouterDecorator } from '@/utils/stories/react-router-decorators';
import { ReactComponent as PlantIcon } from '@public/icons/plant.svg';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof IconButton> = {
  title: 'Components/Button/IconButton',
  component: IconButton,
  decorators: [reactRouterDecorator],
};

export default meta;

type Story = StoryObj<typeof IconButton>;

export const Primary: Story = {
  args: {
    variant: ButtonVariant.primary,
    children: <PlantIcon />,
  },
};

export const Secondary: Story = {
  args: {
    ...Primary.args,
    variant: ButtonVariant.secondary,
  },
};
