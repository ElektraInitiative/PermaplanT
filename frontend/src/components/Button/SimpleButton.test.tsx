import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { ButtonVariant } from '../Button/SimpleButton';
import SimpleButton from '../Button/SimpleButton';

it('renders correctly for primary base variant', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SimpleButton variant={ButtonVariant.primaryBase} />
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly for secondary base variant', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SimpleButton variant={ButtonVariant.secondaryBase} />
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly for primary container variant', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SimpleButton variant={ButtonVariant.primaryContainer} />
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly for secondary container variant', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SimpleButton variant={ButtonVariant.secondaryContainer} />
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
