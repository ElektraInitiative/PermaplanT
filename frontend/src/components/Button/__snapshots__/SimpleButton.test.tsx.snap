// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for primary base variant 1`] = `
<button
  className="disabled:bg-neutral-300 disabled:border-neutral-300 disabled:text-neutral-500 dark:disabled:bg-neutral-300-dark dark:disabled:border-neutral-300-dark dark:disabled:text-neutral-500-dark disabled:cursor-not-allowed button focus:outline-none focus:ring-4 focus:ring-blue-300 border bg-primary-500 dark:bg-primary-300 border-primary-500 hover:bg-primary-600 dark:hover:bg-primary-200 active:bg-primary-900 text-primary-50 dark:text-primary-700"
  type="button"
/>
`;

exports[`renders correctly for primary container variant 1`] = `
<button
  className="disabled:bg-neutral-300 disabled:border-neutral-300 disabled:text-neutral-500 dark:disabled:bg-neutral-300-dark dark:disabled:border-neutral-300-dark dark:disabled:text-neutral-500-dark disabled:cursor-not-allowed button focus:outline-none focus:ring-4 focus:ring-blue-300 border bg-primary-200 dark:bg-primary-600 border-primary-500 hover:bg-primary-200 dark:hover:bg-primary-600 active:bg-primary-900 text-primary-900 dark:text-primary-200"
  type="button"
/>
`;

exports[`renders correctly for secondary base variant 1`] = `
<button
  className="disabled:bg-neutral-300 disabled:border-neutral-300 disabled:text-neutral-500 dark:disabled:bg-neutral-300-dark dark:disabled:border-neutral-300-dark dark:disabled:text-neutral-500-dark disabled:cursor-not-allowed button focus:outline-none focus:ring-4 focus:ring-blue-300 border bg-secondary-500 dark:bg-secondary-300 border-secondary-500 hover:bg-secondary-600 dark:hover:bg-secondary-200 active:bg-secondary-900 text-secondary-50 dark:text-secondary-700"
  type="button"
/>
`;

exports[`renders correctly for secondary container variant 1`] = `
<button
  className="disabled:bg-neutral-300 disabled:border-neutral-300 disabled:text-neutral-500 dark:disabled:bg-neutral-300-dark dark:disabled:border-neutral-300-dark dark:disabled:text-neutral-500-dark disabled:cursor-not-allowed button focus:outline-none focus:ring-4 focus:ring-blue-300 border bg-secondary-200 dark:bg-secondary-600 border-secondary-500 hover:bg-secondary-200 dark:hover:bg-secondary-600 active:bg-secondary-900 text-secondary-900 dark:text-secondary-200"
  type="button"
/>
`;
